subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1447,6 +1821,752 @@ static int tls_mbedtls_init(struct tls_context *context, bool is_server)
+            i++;
+        }
+
+        alpn_buf = XMALLOC(alpn_buf_len, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        if (alpn_buf == NULL) {
+            return -ENOMEM;
+        }
+
+        i = 0;
+        iter = alpn_buf;
+        while(context->options.alpn_list[i] != NULL) {
+            len = XSTRLEN(context->options.alpn_list[i]);
+            XMEMCPY(iter, context->options.alpn_list[i], len);
+            iter += len;
+            *iter++ = ',';
+            i++;
+        }
+
+        /* Replace final trailing comma with NULL terminator */
+        *(iter - 1) = '\0';
+        if (wolfSSL_UseALPN(context->wssl, alpn_buf, alpn_buf_len - 1,
+                WOLFSSL_ALPN_FAILED_ON_MISMATCH) != WOLFSSL_SUCCESS) {
+            XFREE(alpn_buf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            return -EINVAL;
+        }
+
+        XFREE(alpn_buf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    }
+
+    return 0;
+}
+#endif
+
+#if defined(CONFIG_NET_SOCKETS_ENABLE_DTLS)
+static int tls_wolfssl_set_dtls_timeouts(struct tls_context *context)
+{
+    if (context->type == SOCK_DGRAM) {
+        if (wolfSSL_dtls_set_timeout_max(context->wssl,
+                (int)context->options.dtls_handshake_timeout_max) != WOLFSSL_SUCCESS) {
+            return -EINVAL;
+        }
+        if (wolfSSL_dtls_set_timeout_init(context->wssl,
+                (int)context->options.dtls_handshake_timeout_min) != WOLFSSL_SUCCESS) {
+            return -EINVAL;
+        }
+
+        /* Underlying socket always acts like non-blocking due to IO callback overrides */
+        wolfSSL_dtls_set_using_nonblock(context->wssl, 1);
+    }
+
+    return 0;
+}
+#endif
+
+static int tls_wolfssl_set_options(struct tls_context *context)
+{
+    int ret = 0;
+    bool is_server = FALSE;
+
+    if (context->options.role == ZTLS_IS_SERVER) {
+        is_server = TRUE;
+    }
+
+    ret = tls_wolfssl_set_verify(context);
+    if (ret != 0) {
+        return ret;
+    }
+
+    ret = tls_wolfssl_set_hostname(context);
+    if (ret != 0) {
+        return ret;
+    }
+
+    ret = tls_wolfssl_set_session_cache_mode(context);
+    if (ret != 0) {
+        return ret;
+    }
+
+    ret = tls_wolfssl_set_ciphersuites(context);
+    if (ret != 0) {
+        return ret;
+    }
+
+#ifdef HAVE_ALPN
+    ret = tls_wolfssl_set_alpn(context);
+    if (ret != 0) {
+        return ret;
+    }
+#endif
+
+#if defined(CONFIG_NET_SOCKETS_ENABLE_DTLS)
+    ret = tls_wolfssl_set_dtls_timeouts(context);
+    if (ret != 0) {
+        return ret;
+    }
+#endif
+
+    return 0;
+}
+
+static int tls_wolfssl_init(struct tls_context *context, bool is_server)
+{
+    int ret;
+
+    context->options.role = is_server ? ZTLS_IS_SERVER : ZTLS_IS_CLIENT;
+
+#if defined(CONFIG_WOLFSSL_DEBUG)
+    wolfSSL_Debugging_ON();
+#endif
+
+    if (context->type == SOCK_STREAM) {
+        if (is_server) {
+            if (NULL == context->ctx)
+                context->ctx = wolfSSL_CTX_new(wolfTLSv1_2_server_method());
```

Comment:
What about TLS v1.3? Any reason not to use the `wolfSSLv23_server_method()`. It will use the highest available version and allow downgrade. So if you are building with NO_OLD_TLS set (you should) then it will try TLS v1.3 and allow downgrade to TLS v1.2.

Reply:
You can see in protocol_check() that this is specifically a TLS1.2 implementation only. They dont even have defines in net_ip.h for IPPROTO_TLS_1_3. I was thinking we would keep it 1.2 for the first MR to stay consistent and then add 1.3 support after.